File: modules/cobranca.py
"""Módulo de Cobrança e Inadimplência com régua automática."""
import pandas as pd
import streamlit as st

from core import db, ui
from core.slack import enviar_slack

STATUS = ["1º contato", "2º contato", "3º contato", "Pré-negativação", "Regularizado"]
REGUA = [7, 14, 21, 28, 60]


def estagio_regua(dias):
    """Retorna o marco da régua (7/14/21/28/60) apropriado para os dias de atraso."""
    marco = 0
    for d in REGUA:
        if dias >= d:
            marco = d
    return marco


def mensagem_sugerida(nome, dias, valor):
    marco = estagio_regua(dias)
    valor_fmt = f"R$ {valor:.2f}"
    if marco == 0:
        return (f"Olá {nome}! Passando para lembrar da mensalidade de {valor_fmt}. "
                "Qualquer coisa estamos à disposição para ajudar. 💙")
    if marco == 7:
        return (f"Olá {nome}! Identificamos a mensalidade de {valor_fmt} em aberto há {dias} dias. "
                "Conseguimos regularizar hoje? Posso te enviar a 2ª via.")
    if marco == 14:
        return (f"Oi {nome}, tudo bem? Sua pendência de {valor_fmt} está há {dias} dias em aberto. "
                "Podemos combinar uma data para o pagamento? Estou aqui para facilitar.")
    if marco == 21:
        return (f"{nome}, seu débito de {valor_fmt} ({dias} dias) segue em aberto. "
                "Para evitar bloqueios de acesso, vamos resolver juntos? Temos opções de negociação.")
    if marco == 28:
        return (f"{nome}, atenção: sua mensalidade de {valor_fmt} está há {dias} dias vencida. "
                "Precisamos regularizar esta semana para manter sua matrícula ativa.")
    return (f"{nome}, seu débito de {valor_fmt} ultrapassou {dias} dias. Para evitar a "
            "negativação, é essencial regularizar. Vamos negociar condições especiais o quanto antes.")


def render(user):
    ui.titulo("Cobrança e Inadimplência", "Régua de contato 7 · 14 · 21 · 28 · 60 dias e mensagens prontas")
    tab_lista, tab_novo = st.tabs(["📋 Inadimplentes", "➕ Novo registro"])

    with tab_novo:
        with st.form("nova_cobr", clear_on_submit=True):
            c1, c2, c3 = st.columns(3)
            aluno = c1.text_input("Nome do aluno *")
            valor = c2.number_input("Valor em aberto (R$)", 0.0, step=50.0)
            venc = c3.date_input("Data de vencimento")
            c1, c2 = st.columns(2)
            status = c1.selectbox("Status", STATUS)
            resp = c2.text_input("Responsável", value=user["nome"])
            proxima = st.text_input("Próxima ação")
            obs = st.text_area("Observações")
            if st.form_submit_button("Registrar", type="primary"):
                if not aluno:
                    st.error("Informe o nome do aluno.")
                else:
                    db.execute(
                        "INSERT INTO cobrancas (aluno,valor,vencimento,status,ultima_mensagem,"
                        "proxima_acao,responsavel,observacoes,criado_em) VALUES (?,?,?,?,?,?,?,?,?)",
                        (aluno, valor, str(venc), status, "", proxima, resp, obs, db.now()))
                    db.log_action(user["email"], "cobranca_criada", aluno)
                    st.success("Registro criado.")

    with tab_lista:
        c1, c2 = st.columns([2, 3])
        f_status = c1.multiselect("Status", STATUS)
        f_busca = c2.text_input("Buscar aluno")
        rows = db.query("SELECT * FROM cobrancas ORDER BY vencimento")

        data, criticos = [], []
        for r in rows:
            if f_status and r["status"] not in f_status:
                continue
            if f_busca and f_busca.lower() not in (r["aluno"] or "").lower():
                continue
            dias = ui.dias_desde(r["vencimento"]) or 0
            if r["status"] != "Regularizado" and dias >= 60:
                criticos.append((r, dias))
            data.append({
                "ID": r["id"], "Aluno": r["aluno"], "Valor": r["valor"],
                "Vencimento": r["vencimento"], "Dias atraso": dias,
                "Régua": f"{estagio_regua(dias)}d" if dias > 0 else "—",
                "Status": r["status"], "Responsável": r["responsavel"],
            })

        if criticos:
            for r, dias in criticos:
                st.error(f"🚨 **{r['aluno']}** está há **{dias} dias** em atraso (R$ {r['valor']:.2f}).")

        df = pd.DataFrame(data)
        st.caption(f"{len(df)} registro(s)")
        if not df.empty:
            st.dataframe(df, use_container_width=True, hide_index=True)
            ui.botoes_export(df, "cobrancas")

        if criticos and st.button("📨 Enviar alertas ao Slack (>60 dias)"):
            for r, dias in criticos:
                enviar_slack(f":money_with_wings: *Inadimplência >60 dias* — {r['aluno']} "
                             f"({dias}d, R$ {r['valor']:.2f}) · {r['status']}", "cobranca_60d")
            st.success(f"{len(criticos)} alerta(s) enviados.")

        st.divider()
        st.markdown("##### 💬 Régua & mensagem sugerida")
        ids = [r["id"] for r in rows]
        if ids:
            sel = st.selectbox("Selecione o aluno", ids,
                               format_func=lambda i: next(r["aluno"] for r in rows if r["id"] == i))
            r = db.query_one("SELECT * FROM cobrancas WHERE id=?", (sel,))
            dias = ui.dias_desde(r["vencimento"]) or 0
            msg = mensagem_sugerida(r["aluno"], dias, float(r["valor"] or 0))
            st.info(f"Estágio da régua: **{estagio_regua(dias)} dias** · atraso real: {dias} dias")
            st.text_area("Mensagem pronta (copie e envie)", msg, height=120, key="msg_cobr")
            st.caption("Selecione o texto acima e copie (Ctrl/Cmd+C).")

            with st.form("edit_cobr"):
                c1, c2 = st.columns(2)
                status = c1.selectbox("Atualizar status", STATUS,
                                      index=STATUS.index(r["status"]) if r["status"] in STATUS else 0)
                proxima = c2.text_input("Próxima ação", r["proxima_acao"] or "")
                marcar = st.checkbox("Registrar esta mensagem como 'última enviada'")
                if st.form_submit_button("Salvar", type="primary"):
                    db.execute("UPDATE cobrancas SET status=?,proxima_acao=?,ultima_mensagem=? WHERE id=?",
                               (status, proxima, msg if marcar else r["ultima_mensagem"], sel))
                    db.log_action(user["email"], "cobranca_editada", f"#{sel} {r['aluno']}")
                    st.success("Atualizado.")
                    st.rerun()
            if r["ultima_mensagem"]:
                st.caption(f"Última mensagem registrada: {r['ultima_mensagem'][:120]}…")

File: web/modules/cobranca.js
/* Cobrança e Inadimplência — régua de contato 7·14·21·28·60 dias.
   Banner de críticos (>60d), mensagens prontas por estágio e alerta ao Slack. */
import { html } from "htm/preact";
import { useState } from "preact/hooks";
import { store } from "../lib/store.js";
import { useCollection, useUsers, useSelecao } from "../lib/hooks.js";
import {
  Badge, Table, Modal, Text, Num, DateF, Area, Select, FilterSelect, Assignee, Tabs, BulkBar,
  toast, copiar, ExportButtons, brl, diasDesde, reguaEstagio, mensagemCobranca, DOM,
  aplicarEscopo, podeVerTudo,
} from "../lib/ui.js";
import { alertarSlack } from "../lib/notify.js";

const nowTs = () =>new Date().toISOString().slice(0, 19).replace("T", " ");
const vazia = (u) => ({ aluno: "", valor: 0, vencimento: new Date().toISOString().slice(0, 10), status: DOM.cobr_status[0], ultima_mensagem: "", proxima_acao: "", responsavel: u ? u.nome : "", observacoes: "" });

export function View({ user }) {
  const { rows } = useCollection("cobrancas", { order: "vencimento" });
  const users = useUsers();
  const verTudo = podeVerTudo(user.perfil);
  const [escopo, setEscopo] = useState("minhas");
  const [novo, setNovo] = useState(null);   // objeto em criação
  const [edit, setEdit] = useState(null);   // cobrança em edição
  const [fStatus, setFStatus] = useState("");
  const [fBusca, setFBusca] = useState("");

  const base = aplicarEscopo(rows, user, verTudo ? escopo : "minhas");
  const filtradas = base.filter((r) =>
    (!fStatus || r.status === fStatus) &&
    (!fBusca || (r.aluno || "").toLowerCase().includes(fBusca.toLowerCase())));

  const criticos = filtradas.filter((r) =>r.status !== "Regularizado" && (diasDesde(r.vencimento) || 0) >= 60);

  const salvarNovo = async () => {
    if (!novo.aluno.trim()) return toast("Informe o nome do aluno.", "err");
    const r = await store.insert("cobrancas", { ...novo, criado_em: nowTs() });
    await store.logAction(user.email, "cobranca_criada", r.aluno);
    setNovo(null); toast("Registro criado com sucesso!");
  };

  const enviarAlertas = async () => {
    for (const r of criticos) {
      const dias = diasDesde(r.vencimento) || 0;
      await alertarSlack(`:money_with_wings: *Inadimplência >60 dias* — ${r.aluno} (${dias}d, ${brl(r.valor)}) · ${r.status}`, "cobranca_60d");
    }
    toast(`${criticos.length} alerta(s) enviados ao Slack.`, "warn");
  };

  const { sel, toggle, toggleAll, clear } = useSelecao();
  const bulkRegularizar = async () => {
    const ids = [...sel];
    if (!ids.length) return;
    if (!confirm(`Marcar ${ids.length} cobrança(s) como regularizada(s)?`)) return;
    await store.bulkUpdate("cobrancas", ids, { status: "Regularizado" });
    await store.logAction(user.email, "cobrancas_regularizadas_massa", `${ids.length}`);
    clear(); toast(`${ids.length} cobrança(s) marcada(s) como regularizada(s).`);
  };
  const bulkExcluir = async () => {
    const ids = [...sel];
    if (!confirm(`Excluir ${ids.length} cobrança(s)? Essa ação não pode ser desfeita.`)) return;
    await store.bulkRemove("cobrancas", ids);
    await store.logAction(user.email, "cobrancas_excluidas_massa", `${ids.length}`);
    clear(); toast(`${ids.length} cobrança(s) excluída(s).`);
  };

  const cols = [
    { key: "id", label: "ID" },
    { key: "aluno", label: "Aluno" },
    { key: "valor", label: "Valor", render: (r) =>brl(r.valor), csv: (r) =>brl(r.valor) },
    { key: "vencimento", label: "Vencimento" },
    { key: "atraso", label: "Dias atraso", render: (r) =>diasDesde(r.vencimento) ?? 0, csv: (r) =>diasDesde(r.vencimento) ?? 0 },
    { key: "regua", label: "Régua", render: (r) => { const d = diasDesde(r.vencimento) || 0; return d >0 ? reguaEstagio(d) + "d" : "—"; }, csv: (r) => { const d = diasDesde(r.vencimento) || 0; return d >0 ? reguaEstagio(d) + "d" : "—"; } },
    { key: "status", label: "Status", render: (r) =>Badge(r.status), csv: (r) =>r.status },
    { key: "responsavel", label: "Responsável" },
  ];
  const rowClass = (r) => (r.status !== "Regularizado" && (diasDesde(r.vencimento) || 0) >= 60 ? "overdue" : "");

  return html`
    <h1 class="h1">Cobrança e Inadimplência</h1>
    <p class="sub">Régua de contato 7 · 14 · 21 · 28 · 60 dias e mensagens prontas por estágio.</p>

    ${verTudo ? html`<div style="margin-bottom:14px">${Tabs({ value: escopo, onInput: setEscopo,
      options: [["minhas", "Minhas"], ["todas", "Equipe (todas)"]] })}</div>` : ""}

    <div class="toolbar">
      <button class="btn primary" onClick=${() =>setNovo(vazia(user))}>Novo registro</button>
      ${FilterSelect({ label: "Status", value: fStatus, onInput: setFStatus, options: DOM.cobr_status })}
      <div class="grow"><label>Aluno</label>
        <input placeholder="buscar aluno" value=${fBusca} onInput=${(e) =>setFBusca(e.target.value)}/></div>
      ${ExportButtons({ nome: "cobrancas", columns: cols, rows: filtradas, titulo: "Cobranças" })}
    </div>

    ${criticos.map((r) =>html`<div style="background:#fef2f2;border:1px solid #fecaca;color:#991b1b;padding:8px 14px;border-radius:10px;margin-bottom:8px">
      <b>${r.aluno}</b>está há ${diasDesde(r.vencimento) || 0} dias em atraso (${brl(r.valor)}).</div>`)}
    ${criticos.length ? html`<div style="margin-bottom:12px">
      <button class="btn danger" onClick=${enviarAlertas}>Enviar alertas ao Slack (>60 dias)</button></div>` : ""}

    <div class="count">${filtradas.length} registro(s)${verTudo && escopo === "todas" ? " · equipe" : " · minhas"}</div>
    ${BulkBar({ n: sel.size, onClear: clear, actions: [
      { label: `Marcar regularizada (${sel.size})`, kind: "ok", on: bulkRegularizar },
      { label: `Excluir (${sel.size})`, kind: "danger", on: bulkExcluir },
    ] })}
    ${Table({ columns: cols, rows: filtradas, onRow: (r) =>setEdit({ ...r }), rowClass, sel, onSel: toggle, onSelAll: toggleAll })}

    ${novo ? html`<${Modal} title="Novo registro de cobrança" onClose=${() =>setNovo(null)}
      footer=${html`<button class="btn" onClick=${() =>setNovo(null)}>Cancelar</button>
        <button class="btn primary" onClick=${salvarNovo}>Registrar</button>`}>
      ${Text({ label: "Nome do aluno *", value: novo.aluno, onInput: (v) =>setNovo({ ...novo, aluno: v }) })}
      <div class="row c3">
        ${Num({ label: "Valor em aberto (R$)", value: novo.valor, onInput: (v) =>setNovo({ ...novo, valor: v }) })}
        ${DateF({ label: "Vencimento", value: novo.vencimento, onInput: (v) =>setNovo({ ...novo, vencimento: v }) })}
        ${Select({ label: "Status", value: novo.status, onInput: (v) =>setNovo({ ...novo, status: v }), options: DOM.cobr_status })}
      </div>
      <div class="row c2">
        ${Assignee({ value: novo.responsavel, onInput: (v) =>setNovo({ ...novo, responsavel: v }), users })}
        ${Text({ label: "Próxima ação", value: novo.proxima_acao, onInput: (v) =>setNovo({ ...novo, proxima_acao: v }) })}
      </div>
      ${Area({ label: "Observações", value: novo.observacoes, onInput: (v) =>setNovo({ ...novo, observacoes: v }) })}
    <//>` : ""}

    ${edit ? html`<${EditModal} user=${user} cobr=${edit} onClose=${() =>setEdit(null)}/>` : ""}`;
}

function EditModal({ user, cobr, onClose }) {
  const dias = diasDesde(cobr.vencimento) || 0;
  const msg = mensagemCobranca(cobr.aluno, dias, cobr.valor);
  const [status, setStatus] = useState(cobr.status);
  const [proxima, setProxima] = useState(cobr.proxima_acao || "");
  const [marcar, setMarcar] = useState(false);

  const salvar = async () => {
    await store.update("cobrancas", cobr.id, {
      status, proxima_acao: proxima,
      ultima_mensagem: marcar ? msg : cobr.ultima_mensagem,
    });
    await store.logAction(user.email, "cobranca_editada", `#${cobr.id} ${cobr.aluno}`);
    toast("Cobrança atualizada."); onClose();
  };
  const excluir = async () => {
    if (!confirm("Excluir este registro de cobrança?")) return;
    await store.remove("cobrancas", cobr.id);
    await store.logAction(user.email, "cobranca_excluida", `#${cobr.id}`);
    toast("Registro excluído."); onClose();
  };

  return html`<${Modal} title=${"Cobrança — " + cobr.aluno} size="lg" onClose=${onClose}
    footer=${html`<button class="btn danger" onClick=${excluir}>Excluir</button>
      <button class="btn" onClick=${onClose}>Fechar</button>
      <button class="btn primary" onClick=${salvar}>Salvar alterações</button>`}>
    <div style="background:#eff6ff;border:1px solid #bfdbfe;color:#1e3a8a;padding:8px 14px;border-radius:10px;margin-bottom:12px">Estágio da régua: <b>${reguaEstagio(dias)}</b>dias · atraso real: <b>${dias}</b>dias</div>
    <div class="field"><label>Mensagem pronta (copie e envie)</label>
      <textarea class="copybox" rows=${5} readonly>${msg}</textarea></div>
    <div style="margin-bottom:12px"><button class="btn" onClick=${() =>copiar(msg)}>Copiar mensagem</button></div>
    <div class="row c2">
      ${Select({ label: "Atualizar status", value: status, onInput: setStatus, options: DOM.cobr_status })}
      ${Text({ label: "Próxima ação", value: proxima, onInput: setProxima })}
    </div>
    <label class="field" style="flex-direction:row;align-items:center;gap:8px">
      <input type="checkbox" checked=${marcar} onChange=${(e) =>setMarcar(e.target.checked)}/>
      <span>Registrar esta mensagem como última enviada</span></label>
    ${cobr.ultima_mensagem ? html`<div class="section-t">Última mensagem registrada</div>
      <div class="hist"><div class="h-item">${cobr.ultima_mensagem}</div></div>` : ""}
  <//>`;
}
